### ISSUE TITLE
build error on Apple MacOs with M3

### ISSUE THREAD

**Issue by noondie** [MartinHelmut/cpp-gui-template-sdl2#24]
Hello 
first I want to say a big thank you for your article and all those explanations. 

I tried to compile your code as described in the QuickStart doc. 

#cmake -GXcode -DCMAKE_BUILD_TYPE=Release -B build/Xcode

and 
cmake --build build/xcode 

and got the following error 
clang: error: no such file or directory: 'x86_64/TestRunner.o'
which is correct ;=) 

where can I specify the correct platform? And why is the platform wrong estimated? 

thanks for your help

**Comment by noondie** [comment]
add some more info


Ld /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Debug/Objects-normal/arm64/Binary/ResourcesTest normal arm64 (in target 'ResourcesTest' from project 'BasicGuiProjectSetup')
    cd /Users/conrad/Sources/cpp-gui-template-sdl2
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -Xlinker -reproducible -target arm64-apple-macos14.3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -O0 -L/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/EagerLinkingTBDs/Debug -L/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/core/Tests/Debug -F/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/EagerLinkingTBDs/Debug -F/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/core/Tests/Debug -filelist /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Debug/Objects-normal/arm64/ResourcesTest.LinkFileList -Xlinker -rpath -Xlinker /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/sdl2-build/Debug -Xlinker -object_path_lto -Xlinker /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Debug/Objects-normal/arm64/ResourcesTest_lto.o -Xlinker -no_deduplicate -Wl,-search_paths_first -Wl,-headerpad_max_install_names /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/TestRunner.build/Debug/Objects-normal/arm64 x86_64/TestRunner.o /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/core/Debug/libCore.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/spdlog-build/Debug/libspdlogd.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/fmt-build/Debug/libfmtd.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/vendor/imgui-setup/Debug/libimgui.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/sdl2-build/Debug/libSDL2-2.0d.0.dylib /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/settings/Debug/libSettings.a -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Debug/Objects-normal/arm64/ResourcesTest_dependency_info.dat -o /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Debug/Objects-normal/arm64/Binary/ResourcesTest
clang: error: no such file or directory: 'x86_64/TestRunner.o'

**Comment by noondie** [comment]
in the next step I tried 
% cmake  -DCMAKE_BUILD_TYPE=Release -B build/xcode
% cmake --build build/xcode
which compiled fine 

after that 
% cpack --config build/xcode/CPackConfig.cmake  

which build the dmg file , but the icon in the install process is broken. 
I guess this is because after installation the App into the Application folder,  I get an error when executing. 
"You cannot open the Program App since it is damaged or incomplete"
 
this is the folder struct 

% pwd
/Applications/App.app
% tree
.
└── Contents
    ├── Frameworks
    │   └── libSDL2-2.0.0.dylib
    ├── Info.plist
    ├── MacOS
    │   └── App
    └── Resources
        ├── Manrope.ttf
        └── icon.icns

if I execute App, the app start running.

**Comment by MartinHelmut** [comment]
Hey and thanks for reporting! I'll try to replicate what you have. Do I assume right you did all over the a terminal, e.g. iTerm? Could you also post your CMake, XCode, and macOS version for me as reference?

**Comment by MartinHelmut** [comment]
Thanks for adding the used versions! I was able to reproduce the issue and [created a first PR](https://github.com/MartinHelmut/cpp-gui-template-sdl2/pull/25/files) where I try to solve the problem. One part is the defined architecture. I reworked this part for macOS to be defined more explicit.

The current **TL;DR** is:

- Changed `CMAKE_OSX_ARCHITECTURES` to be a cache variable
- Specified `x86_64;arm64` (universal executable) as default value for presets

***

**Long form:**

By having the `CMAKE_OSX_ARCHITECTURES` variable a cache variable it can be overridden on command usage with the desired target Apple architecture. For example, for a universal executable the value `x86_64;arm64` should be used. Making the command a bit longer, but more explicit:

```shell
cmake -GXcode -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES="x86_64;arm64" -B build/xcode
cmake --build build/xcode
```

The issue with this is I cannot run CPack yet![^1] Still investigating, because it works with presets (read further).

I also defined creating universal executables via presets. This gives a very convenient command I'll probably take into the quick start guide documentation.[^2]

```shell
cmake --preset xcode-release
cmake --build --preset xcode-release
```

The actual distribution can then be created without problems (at least for me) with the defined Xcode CPack preset:

```shell
cpack --preset xcode-release
```

Or, thanks to CMake workflows, all of this, after a fresh clone, works with one command:

```shell
cmake --workflow --preset xcode-dist
```

**Let me know** if you can run the preset commands, e.g. the workflow version of it (after removing the `build` folder first for a fresh run).

[^1]: For some reason when I run the commands separately, not as presets, even with `CMAKE_BUILD_TYPE` defined as `Release`, I get always a `Debug` folder from Xcode.
[^2]: I will also need to update all other template repos and the article 😅

**Comment by MartinHelmut** [comment]
Found out why CPack with the base command did not work, but with the Preset did: For Xcode to pick a _Release_ configuration it is necessary to pass `--config Release` on build, like this:

```shell
cmake -GXcode -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES="x86_64;arm64" -B build/xcode
cmake --build build/xcode -- config Release
```

Only then can CPack be run successfully, finding the right folder for the release distribution:

```shell
cpack --config build/xcode/CPackConfig.cmake
```

I also updated the PR: https://github.com/MartinHelmut/cpp-gui-template-sdl2/pull/25 I plan to merge the PR only after I tested some more scenarios in Windows and Linux as well, just to be sure how this option and the usage of presets can influence results on the other platforms.

**Comment by noondie** [comment]
fine, thanks a lot for the expanations. 

One remaining question: Is this a universal build for Mac, Does the dmg run on both platforms by the definition of the ARCH in the command line? 

regards c~b

**Comment by MartinHelmut** [comment]
Hey, yes, as I wrote:

> for a universal executable the value `x86_64;arm64` should be used ...

So either using this on the command line, or using the preset command where this is the default. 

Does everything run now for you?

**Comment by noondie** [comment]
Hello 

yes both solutions are working. 
i will try this Universal build on the other machine in the next days. 

Unfortunately i have to disable the test build otherwise i will get an compile error 
I will check this in the next days in more detail. 

Do you have a hint how to create a subfolder structure in the Resource folder , i.e. one for pics, one for fonts , one for texts etc?

Regards c~b

**Comment by MartinHelmut** [comment]
> Unfortunately i have to disable the test build otherwise i will get an compile error

Can you post some more details about this? E.g. the error output? Would love to see what the issue is with that :)

**Regarding subfolder:** I think this should be possible via [MACOSX_PACKAGE_LOCATION](https://cmake.org/cmake/help/latest/prop_sf/MACOSX_PACKAGE_LOCATION.html) as [used in the AppAsset.cmake file](https://github.com/MartinHelmut/cpp-gui-template-sdl2/blob/main/src/app/cmake/AppAssets.cmake#L16). Here you could probably define the set of source files you want to differentiate and set the target property on them.

```cmake
# Split different types of assets
set(FONT_ASSETS ${PROJECT_SOURCE_DIR}/src/assets/fonts/Manrope.ttf)
set(IMAGE_ASSETS
  ${PROJECT_SOURCE_DIR}/src/assets/images/img1.png
  ${PROJECT_SOURCE_DIR}/src/assets/images/img2.png)

# Platform specific static assets
if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
  # ...
elseif (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
  set(MACOSX_ICON_ASSET ${PROJECT_SOURCE_DIR}/src/assets/icons/icon.icns)
  set(ALL_MACOSX_STATIC_ASSETS
    ${FONT_ASSETS}
    ${IMAGE_ASSETS}
    ${MACOSX_ICON_ASSET})

  # Define subfolders here:
  set_source_files_properties(${MACOSX_ICON_ASSET}
    PROPERTIES MACOSX_PACKAGE_LOCATION ${CMAKE_INSTALL_DATADIR})
  set_source_files_properties(${FONT_ASSETS}
    PROPERTIES MACOSX_PACKAGE_LOCATION ${CMAKE_INSTALL_DATADIR}/fonts)
  set_source_files_properties(${IMAGE_ASSETS}
    PROPERTIES MACOSX_PACKAGE_LOCATION ${CMAKE_INSTALL_DATADIR}/images)

  target_sources(${NAME} PUBLIC ${ALL_MACOSX_STATIC_ASSETS})
elseif (CMAKE_SYSTEM_NAME STREQUAL "Linux")
  # ...
endif ()
```

**Remark:** I did not test this, but I think this is how it should work.

**Comment by noondie** [comment]
thanks for your help and sorry for the delay. i was a little bit busy ;=)  

the hint with the subfolder have the same result as my trials 

i changed the AppAsset make file and got the following result 

% tree
.
├── Frameworks
│   └── libSDL2-2.0.0.dylib
├── Info.plist
├── MacOS
│   └── App
├── PkgInfo
└── share
    ├── fonts
    │   └── Manrope.ttf
    ├── icon.icns
    └── images
        ├── img1.png
        └── img2.png


as you see, there is a share folder. Even the icon.icns cannot be found be the app and is not shown. 


Anyway, the second problem with your test files is the following 


% git clone https://github.com/MartinHelmut/cpp-gui-template-sdl2.git
% cd cpp-gui-template-sdl2
% cmake -GXcode -DCMAKE_BUILD_TYPE=Release -B build/xcode
% cmake --build build/xcode --config Release


————  
 /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Release/Objects-normal/arm64/Binary/ResourcesTest normal arm64 (in target 'ResourcesTest' from project 'BasicGuiProjectSetup')     cd /Users/conrad/Sources/cpp-gui-template-sdl2    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -Xlinker -reproducible -target arm64-apple-macos10.15 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk -O3 -L/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/EagerLinkingTBDs/Release -L/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/core/Tests/Release -F/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/EagerLinkingTBDs/Release -F/Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/core/Tests/Release -filelist /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Release/Objects-normal/arm64/ResourcesTest.LinkFileList -Xlinker -rpath -Xlinker /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/sdl2-build/Release -Wl,-search_paths_first -Wl,-headerpad_max_install_names /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/TestRunner.build/Release/Objects-normal/arm64 x86_64/TestRunner.o /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/core/Release/libCore.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/spdlog-build/Release/libspdlog.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/fmt-build/Release/libfmt.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/vendor/imgui-setup/Release/libimgui.a /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/_deps/sdl2-build/Release/libSDL2-2.0.0.dylib /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/src/settings/Release/libSettings.a -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Release/Objects-normal/arm64/ResourcesTest_dependency_info.dat -o /Users/conrad/Sources/cpp-gui-template-sdl2/build/xcode/build/ResourcesTest.build/Release/Objects-normal/arm64/Binary/ResourcesTest
clang: error: no such file or directory: 'x86_64/TestRunner.o'

**Comment by MartinHelmut** [comment]
Hey 👋🏻

## Issue 1 - Build issue

Let's start with this one. The build error I can reproduce, reason is, as [described in the docs](https://github.com/MartinHelmut/cpp-gui-template-sdl2/blob/main/docs/QuickStart.md#build), when not using [CMake Presets](https://github.com/MartinHelmut/cpp-gui-template-sdl2/blob/main/docs/CMakePresets.md) the target architecture needs to be set explicitly via `CMAKE_OSX_ARCHITECTURES` on macOS, e.g. for universal builds to `x86_64;arm64`:

```shell
cmake -GXcode -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES="x86_64;arm64" -B build/xcode
cmake --build build/xcode --config Release
```

## Issue 2 - Bundle structure

I actually cannot reproduce your folder structure when running CPack. One thing that is very strange to me:

> as you see, there is a share folder

A share folder should only be created at Windows or Linux, not macOS, where it should be a `Resources` folder. So this is really confusing to me, as if the environment is wrongly detected by CMake 🤔 Could you give me the full log for all three commands ([ref](https://github.com/MartinHelmut/cpp-gui-template-sdl2/blob/main/docs/Packaging.md#macos-1)) to create a bundle?

```shell
cmake -GXcode -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_ARCHITECTURES="x86_64;arm64" -B build/xcode
cmake --build build/xcode --config Release
cpack --config build/xcode/CPackConfig.cmake
```

This should help me figuring out what is going on on your end.

**Comment by noondie** [comment]
Hello MArtin, 

thanks . 
Issue 1: works with the mentioned option 
Issue 2: 
i add the output as attachment 
i guess, this is the problem 

set_source_files_properties(/Users/conrad/Sources/cpp-gui-template-sdl2_/src/assets/images/img1.png;/Users/conrad/Sources/cpp-gui-template-sdl2_/src/assets/images/img2.png PROPERTIES MACOSX_PACKAGE_LOCATION share/images )

PACKAGE_LOCATION is wrong. What do you think 

regards c~b

￼

**Comment by MartinHelmut** [comment]
Ahh yes I see, the package location needs to specify the installation data directory by using `CMAKE_INSTALL_DATADIR`:

```cmake
set_source_files_properties("<files ...>"
  PROPERTIES MACOSX_PACKAGE_LOCATION ${CMAKE_INSTALL_DATADIR}/images)
```

So instead of using `share/images` it would be `${CMAKE_INSTALL_DATADIR}/images` to create a `Resources/images` folder in the app bundle.

**Comment by noondie** [comment]
hmm, 
this doesn’t work on my side. 

what work for me is to reconfigure 
CMAKE_INSTALL_DATADIR since this contains share. 
But if you set it before the call of set_source_files_properties something like
set (CMAKE_INSTALL_DATADIR "Resources")

it works well. What do you think…

**Comment by MartinHelmut** [comment]
Alright, I had to understand what is going on 😅 But, I think I figured it out now. I'll start with a short version.

## TL;DR

No need to `set(CMAKE_INSTALL_DATADIR "Resources")`. Two things needed, first, remove the `RESOURCE` property from the target properties for macOS:

```diff
set_target_properties(${NAME} PROPERTIES
  XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY ""
  MACOSX_BUNDLE_BUNDLE_VERSION "${BUILD_VERSION}"
  MACOSX_BUNDLE_SHORT_VERSION_STRING "${PROJECT_VERSION}"
  MACOSX_BUNDLE_INFO_PLIST "${CMAKE_CURRENT_SOURCE_DIR}/Manifests/Info.plist"
  MACOSX_BUNDLE_GUI_IDENTIFIER "${PROJECT_COMPANY_NAMESPACE}.${CMAKE_PROJECT_NAME}"
  MACOSX_BUNDLE_COPYRIGHT "(c) ${CURR_YEAR} ${PROJECT_COMPANY_NAME}"
  INSTALL_RPATH @executable_path/../Frameworks
-  RESOURCE "${MACOSX_STATIC_ASSETS}")
+ )
```

Second, do not use `CMAKE_INSTALL_DATADIR` and define the folder directly, as relative path (as it is relative to the app bundle):

```cmake
set_source_files_properties("<files ...>"
  PROPERTIES MACOSX_PACKAGE_LOCATION "Resources/images")
```

## Long explanation

I was able to reproduce the same behaviour, getting a `share` folder. But **only** under the condition I remove the `RESOURCE` property from `set_target_properties` in `src/app/cmake/packaging/Darwin.cmake`:

```diff
set_target_properties(${NAME} PROPERTIES
  XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY ""
  MACOSX_BUNDLE_BUNDLE_VERSION "${BUILD_VERSION}"
  MACOSX_BUNDLE_SHORT_VERSION_STRING "${PROJECT_VERSION}"
  MACOSX_BUNDLE_INFO_PLIST "${CMAKE_CURRENT_SOURCE_DIR}/Manifests/Info.plist"
  MACOSX_BUNDLE_GUI_IDENTIFIER "${PROJECT_COMPANY_NAMESPACE}.${CMAKE_PROJECT_NAME}"
  MACOSX_BUNDLE_COPYRIGHT "(c) ${CURR_YEAR} ${PROJECT_COMPANY_NAME}"
  INSTALL_RPATH @executable_path/../Frameworks
-  RESOURCE "${MACOSX_STATIC_ASSETS}")
+ )
```

Removing this property yielded the `share` folder using any version of:

```cmake
set_source_files_properties("<files ...>"
  PROPERTIES MACOSX_PACKAGE_LOCATION ${CMAKE_INSTALL_DATADIR}/images)
```

Looking at the [documentation for GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html), this should have not surprised me. As `DATADIR` points to `DATAROOTDIR` that is indeed the value `share`. This only never became an issue for me as the `RESOURCE` property moved assets into the `Resources` folder in a flat structure. Also, `CMAKE_INSTALL_DATADIR` is a cache variable. If it was set to a different value it will keep the value, even if the `set` call is removed later. It only changes then when **removing the `build` folder** proper, and re-running the commands.

The problem then is, it is either the `RESOURCE` property for the macOS target, or setting file properties directly via `set_source_files_properties` to actually create a dedicated folder structures. Doing this, removing the `RESOURCE` property in `src/app/cmake/packaging/Darwin.cmake`, and then using `set_source_files_properties` with a relative path, will give the desired result. Important is then to **not use the** `CMAKE_INSTALL_DATADIR`, but define `Resources` directly.

Picking up the earlier example this could look like this:

```cmake
# Split different types of assets
set(FONT_ASSETS ${PROJECT_SOURCE_DIR}/src/assets/fonts/Manrope.ttf)
set(IMAGE_ASSETS
  ${PROJECT_SOURCE_DIR}/src/assets/images/img1.png
  ${PROJECT_SOURCE_DIR}/src/assets/images/img2.png)

# Platform specific static assets
if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
  # ...
elseif (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
  set(MACOSX_ICON_ASSET ${PROJECT_SOURCE_DIR}/src/assets/icons/icon.icns)
  set(ALL_MACOSX_STATIC_ASSETS
    ${FONT_ASSETS}
    ${IMAGE_ASSETS}
    ${MACOSX_ICON_ASSET})

  # Define subfolders here:
  set_source_files_properties(${MACOSX_ICON_ASSET}
    PROPERTIES MACOSX_PACKAGE_LOCATION "Resources")
  set_source_files_properties(${FONT_ASSETS}
    PROPERTIES MACOSX_PACKAGE_LOCATION "Resources/fonts")
  set_source_files_properties(${IMAGE_ASSETS}
    PROPERTIES MACOSX_PACKAGE_LOCATION "Resources/images")

  target_sources(${NAME} PUBLIC ${ALL_MACOSX_STATIC_ASSETS})
elseif (CMAKE_SYSTEM_NAME STREQUAL "Linux")
  # ...
endif ()
```

Small note, if changing the folder for the font this also needs to be adjusted in `src/core/Core/Application.cpp`:

```diff
- const std::string font_path{Resources::font_path("Manrope.ttf").generic_string()};
+ const std::string font_path{Resources::font_path("fonts/Manrope.ttf").generic_string()};
```

**Comment by noondie** [comment]
Hello , 

yes, i see the trick. 
cmake is really a tough guy. 

Thanks for all your Effort!

sizu c~b


> Am 15.04.2024 um 22:16 schrieb Martin Helmut Fieber ***@***.***>:
> 
>

**Comment by MartinHelmut** [comment]
> cmake is really a tough guy

It is indeed 😅 But also fun to tackle! Glad I was able to help. In that case I would close this issue now as resolved ✌🏻 Don't hesitate to reach out again, I can also always reopen this issue if there comes up anything else.

**Comment by noondie** [comment]
Hello 

yes, fine with closing. 

One question came to my mind. Have you ever tried to do this with iOS? 
Redards  c~b

**Comment by MartinHelmut** [comment]
> Have you ever tried to do this with iOS?

Yes indeed 😅 I'm literally currently (last week's) working on getting the template ready for mobile devices in general. I started with [SDL3 version](https://github.com/MartinHelmut/cpp-gui-template-sdl3) and am testing it currently. After that I will port it to this and other templates too.

**Comment by noondie** [comment]
👍

let me know if you have published your code…
